server/src/main/java/org/elasticsearch/action/search/SearchProgressListener.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.lucene.search.TotalHits;
+import org.elasticsearch.cluster.routing.GroupShardsIterator;
+import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.search.aggregations.InternalAggregations;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+/**
+ * A listener that allows to track progress of the {@link SearchAction}.
+ */
+abstract class SearchProgressListener {
+    private static final Logger logger = LogManager.getLogger(SearchProgressListener.class);
+
+    public static final SearchProgressListener NOOP = new SearchProgressListener() {};
+
+    private List<SearchShard> shards;
```

Comment:
this list is mutable, it's only set in notifyListShards , I wonder if we are good in terms of thread-safety and visibillity from other threads? Given that the instance member is only used for error messages as part of log warnings, wouldn't it be better that the methods where it's used already get the complete shard id instead of having to look it up? Maybe too complex to implement?